Less code, more docs

thed636 opened the pull request:
Remove some code and add the very first version of How To.

Commit:
Remove redundant send_impl and recv_impl specializations

While writing docs made some changes which lead to a decrease in LOC.
This is a little, but good commit to remove unnecessary code.

Commit:
Add very handy (I hope so) shortcuts for user

Commit:
Add basic How To - not ideal, but better than nothing

Commit:
Add compilation flags export

Useful for doxygen/doxypress and other documentation generators

Commit:
Remove is_pure_connection and connection_traits

Both are not needed.

elsid commented on README.md:
```diff
@@ -10,7 +10,7 @@ The library leverages the power of template metaprogramming, providing convenien
 ### API
 
-Since the project is on early state of developmet it lacks of documentation. We understand the importance of good docs and are working hard on this problem. But until the work isn't done there are [unit tests](tests/request_integration.cpp) which can help you understand the main use-cases of OZO on this early state of the project. We also have a C++Now'18 talk about OZO. The persentation is [here ...
+Since the project is on early state of developmet it lacks of documentation. We understand the importance of good docs and are working hard on this problem. Complete documentation is on the way, but now: look at our [HOW TO](docs/howto.md), then [unit tests](tests/request_integration.cpp) which can help you understand the main use-cases of OZO. We also have a C++Now'18 talk about OZO. The persent ...
```
Let's split this paragraph by lines of ~100 symbols.

elsid replied:
`developmet` -> `development`
`Video will be provided later at` it is already there.

thed636 replied:
**Done**

elsid commented on docs/howto.md:
```diff
@@ -0,0 +1,174 @@
+    ozo::connection_info<> conn_info("host=... port=...");
+
+    // Our query statement
+    const auto query = "SELECT id, name FROM users_info WHERE amount>="_SQL + std::int64_t(25);
```
Add `using namespace ozo::literals`.

thed636 replied:
**Done**

elsid commented on docs/howto.md:
````diff
@@ -0,0 +1,174 @@
+ozo::rows_of<std::int64_t, std::optional<std::string>> rows;
+```
+
+Here we define result type as we want to see it. Practically `ozo::rows_of` is an alias on `std::vector<std::tuple<...>>`. And `ozo::into` is an alias on `std::back_inserter`. So `request()` function will fill this vector of tuples according to the database response rows. It is very important to preserve the same order of fields in request and types in the tuple (it is a little bit annoyng, but t ...
````
Split this paragraph by lines of ~100 symbols.

elsid replied:
`annoyng` -> `annoying`

thed636 replied:
**Done**

elsid commented on docs/howto.md:
````diff
@@ -0,0 +1,174 @@
+OZO_PG_DEFINE_TYPE_AND_ARRAY(boost::uuids::uuid, "uuid", UUIDOID, 2951, bytes<16>)
+```
+
+`OZO_PG_DEFINE_TYPE` defines a C++ to built-in PostgreSQL type mapping. First argument here is a C++ type, second - PostgreSQL type name. Fird is type OID in PostgreSQL, fourth - type size which can be static for N bytes `bytes(N)`, or dynamic `dynamic_size`, like in this case.
````
`Fird` -> `First`

thed636 replied:
**Done**

Commit:
Remove ConnectionWrapper concept

It is useless now because in practice it is an alias to Nullable.

Commit:
Basic Doxygen/Doxypress style in source documentation

Commit:
Change unwrap_connection implementation to structures

Commit:
Add native handles doxydocuments

Commit:
Add concepts.h doxydocuments

Commit:
Add connection.h doxydocuments

Commit:
Replace Require<> with static_assert where possible

Require<> is good thing for template overloading, but it is
not convenient for mandatory requirements of the function.
It is better to use static_assert where possible.

Commit:
Move get_result_status_name to the separate file

Now it can be ignored by code coverage, since no code coverage is needed here.

elsid commented on include/ozo/connection.h:
```diff
@@ -12,20 +12,13 @@ namespace ozo {
 template <typename T>
 inline decltype(auto) unwrap_connection(T&& conn,
-        Require<!ConnectionWrapper<T>>* = 0) noexcept {
+        Require<!Nullable<T>>* = 0) noexcept {
```
`std::reference_wrapper` is not nullable but `unwrap_connection` is required to get value from it.

thed636 replied:
Well, this version of `unwrap_connection` does not work with `std::reference_wrapper` so as it didn't work in the previous version too. So nothing has changed. To support `std::reference_wrapper` the overloaded version of `unwrap_connection` is needed. Something like this (according to one of the next commits):
```c++
template <typename T>
struct unwrap_connection_impl<std::reference_wrapper<T>> {
    static constexpr decltype(auto) apply(std::reference_wrapper<T> ref) noexcept {
        return unwrap_connection(ref.get());
    }
};
```
It has to be done thoroughly but in a different pull request.

elsid commented on include/ozo/connection.h:
```diff
@@ -63,6 +134,32 @@ struct get_connection_socket_impl {
+ * @brief Get the connection socket object
+ *
+ * Connection socket object getter. Connection must provide socket for IO. In general it must return
+ * reference to an instance of [boost::asio::posix::stream_descriptor](https://www.boost.org/doc/libs/1_67_0/doc/html/boost_asio/reference/posix__stream_descriptor.html).
```
We support boost 1.66 in CI, better use links to this version.

thed636 replied:
**Done**

elsid commented on include/ozo/connection.h:
```diff
@@ -91,6 +214,34 @@ struct get_connection_error_context_impl {
+ * Connection OZO error context getter. In many cases the `error_code` is a good thing, but sometimes
+ * you need an additional context which unhopefully can not be placed inside an `error_code` object. To
+ * solve this problem this additional error context is introduced. Usually it returns a reference to an
+ * std::string object. The error context is reseted inside get_connection implementation for Connection.
```
reseted -> reset

thed636 replied:
**Done**

elsid commented on include/ozo/connection.h:
```diff
@@ -344,20 +680,86 @@ struct is_connection_provider<T, std::void_t<decltype(
+    void (error_code ec, connection_type<ConnectionProvider> conn);
+ * @endcode
+ *
+ * This behavour can be customized via async_get_connection_impl specialization. E.g. for Connection
```
behavour -> behaviour
Use spelling checker, Luke!

thed636 replied:
**Done**

Thanks, I will.

Commit:
Add Doxypress-generated docs
